### PR TITLE
Textclust

### PR DESCRIPTION
**Pull request by Dennis1989**
Integration of a text-based stream clustering algorithm (textClust) into river, as discussed here: https://github.com/online-ml/river/discussions/1006

### CODE REVIEW

**Comment by MaxHalford** [comment]
Thanks for starting a PR Dennis. There are quite a few style issues to fix, but nothing we can't solve together. Do you mind if I nit-pick and add a lot comments? 🤗

We're going to release a new version of River sometime this week or the next. I think it's better to take our time with this PR, and include it in the next version River (we make a new release roughly every 3 months, sometimes less). In that regard, we can take our time with this PR :)

**Comment by Dennis1989** [comment]
No worries Max, go ahead. I am not really a software engineer but am certainly interested in fixing everything that is not adequate.

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:13]: It's more common in Python to use `PascalCase` for class names. So I would suggest calling this `TextClust`. Else some people are going to think this is a function.

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:6]: An issue here is that scikit-learn is not one of River's dependencies, and we would like to keep it like that. Agglomerative clustering is not a difficult algorithm to implement, so maybe we could implement a version ourselves. Also that might work better with the fact that we're using dictionaries.

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:18]:
```suggestion
    r"""textClust is a clustering algorithm designed for text.

    textClust[^1][^2] is a stream clustering algorithm for textual data that can identify and track topics over time in a stream of texts.
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:22]: The line lengths are a bit long here. We use a maximum of 100. Normally black fixes this automatically, but not for comments I believe (you have to do it yourself).

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:27]:
```suggestion
        Distance threshold to merge two micro-clusters. Must be within the range `(0, 1]`
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:28]: It's preferable not to have parameters with a leading `_`. But of course `lambda` is a reserved keyword. In this case maybe you can use `fading_factor`.

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:32]: `term_fading` is preferable

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:36]:
```suggestion
    micro_distance
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:34]: `real_time_fading` is preferable

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:53]: `micro_clusters` is preferable

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:84]:
```suggestion
    ...     feature_extraction.BagOfWords(lowercase=True, ngram_range=(1, 2), stop_words=stopwords),
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:92]:
```suggestion
    ...     model = model.learn_one(x["text"], id=x["idd"])
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:130]: `last_cleanup` is preferable

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:137]: `micro_to_macro` is preferable

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:138]: `up_to_date` is preferable

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:140]: Shouldn't a lot of these attributes be private? I don't think you want to expose to users. If such is the case, I would add a leading `_` to mark them as private.

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:146]: I would call the `time` parameter `t`

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:208]: This happens if the input dictionary is empty right? I would ignore such a case, it's up to the users (or some other class) to have this responsibility of checking inputs. Also, you're not supposed to `print` errors, but to `raise` them

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:218]: You have to document the `type` parameter here. Also wouldn't it be preferable to have this in the `__init__`?

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:265]:
```suggestion
                threshold = mu - self.sigma * math.sqrt(squaresum / (counter - 1) - mu**2)
```

**Review by MaxHalford — COMMENTED** [review]
Here's a first review pass :)

There's still a lot of stuff I would change, but this should get the ball rolling. Feel free to push back and ask for questions if I haven't been clear enough.

Overall, I'm ok merging this PR if we sort out the `sklearn.AgglomerativeClustering` issue.

Once this is merged, I don't mind taking some time to clean up the code myself. We have a bit of a coding style at River and we want to make sure all our code is somewhat homogeneous. It's a lot to ask of you to adapt to this style, so I really don't mind jumping in and doing the necessary changes myself (with your authorization of course).

**Comment by hoanganhngo610** [comment]
Dear @Dennis1989! I see that this PR is inactive for quite some time. Would you mind if I take over and finish what we're currently having? It seems like most of them would be code-style related, and a fresh implementation of Agglomerative Clustering.

**Comment by Dennis1989** [comment]
@hoanganhngo610 @MaxHalford Thank you both for your suggestions and support. I was on vacation the last two weeks and just returned yesterday. This week I plan to incorporate the suggested changes (which I think are all valid and necessary). If I encounter problems, I will return to both of you, if this is okay.

File: river/cluster/textclust.py
Comment (Dennis1989) [reply river/cluster/textclust.py:6]: The question here is, where do we place this implementation? Utils? Also, in what detail should this be implemented: 

1) Textclust provides a pre-computed distance matrix. It is possible to implement a version that only uses this mode. Otherwise, it would be possible to additionally allow different distance metrics (e.g., euclidean, cosine, etc.). 

2) Which linkage modes should we offer (single, complete average?)

Overall we should decide if we only treat agglomerative clustering as util for textclust or if we should implement it with the intention to be used by other algorithms as well.

**Comment by Dennis1989** [comment]
I think I addressed most of the issues mentioned by @MaxHalford. Feel free to iterate over the code again, and don't hold back with criticism :)

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:16]:
```suggestion
    r"""textClust, a clustering algorithm for text data. 

    textClust[^1][^2] is a stream clustering algorithm for textual data that can identify and track topics
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:19]:
```suggestion
    approach where the stream is first summarised in real-time.
    
    The result is many small preliminary clusters in the stream called `micro-clusters`.
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:163]: You have to document this behavior somewhere for users to understand.

I think we should look into consistently having an `id` parameter in the `learn_one` method of every clustering method (cc @hoanganhngo610)

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:246]:
```suggestion
    def _get_closest_mc(self, mc, idf, distance):
        """Return the closest micro cluster."""
```

File: river/cluster/textclust.py
Comment (MaxHalford) [inline river/cluster/textclust.py:388]: It's not efficient to call `.loc` many times like this. You should instead use simply Python lists, or a numpy array. And then convert the data to a pandas DataFrame. But I suppose we don't have to look into performance in this PR :)

**Review by MaxHalford — APPROVED** [review]
Here's some more comments.

There's a few thing I would still change, such as using dataclasses. But I think we can merge this as it is, so you can start working on the example notebook, and then I can do some refactoring work on the side if that's ok with you. I'll make a pull request to show you what I would do, and we can agree/disagree :)

File: river/cluster/textclust.py
Comment (Dennis1989) [reply river/cluster/textclust.py:388]: I agree. Not super efficient, but we should keep in mind that this method is only called when macro-clusters are generated from micro-clusters (which usually does not happen during the online phase). Therefore it is (a) really infrequently executed and (b) only used on the significantly smaller number of micro clusters.

Comment (MaxHalford) [reply river/cluster/textclust.py:388]: Understood. We could run some profiling to understand what could be improved anyway. But it's not important right now.

File: river/cluster/textclust.py
Comment (Dennis1989) [reply river/cluster/textclust.py:16]: So do I need to fix your comments first or how do we proceed?

Comment (MaxHalford) [reply river/cluster/textclust.py:16]: Yes you can fix these first I would say :)

File: river/cluster/textclust.py
Comment (hoanganhngo610) [reply river/cluster/textclust.py:163]: @MaxHalford  I also agree, since CluStream also depends a lot on distinguishing different micro clusters with unique IDs. My main concern is that how do we choose the format for the ID when two micro-clusters are formed?

Comment (MaxHalford) [reply river/cluster/textclust.py:163]: I don't know, I'll have to look into it. Let's not mix sample IDs with cluster IDs, I'm not sure which ones you're talking about

Comment (Dennis1989) [reply river/cluster/textclust.py:163]: Yes, @MaxHalford is right. These are two different things. 

So why do we need IDs for observations? Especially for text-based input data, a lot of metadata is associated (thinking of Tweets that hold user objects, locations, and other important information). This information can be used within post hoc analyses to find additional characteristics of clusters that are only built based on the text (e.g., which users belong to a certain cluster). Therefore it is necessary to store the IDs. The question here is if IDs should also be subject to concept drift to understand the current importance of observations. What I think is super important is that we do not "fade away" cluster IDs. It should always be possible to know which ID belongs to which cluster. However, we also have to consider what happens if a cluster itself is removed from the current clustering (maybe cluster - ID assignment should be persisted in secondary memory?)

So why do we need IDs for clusters? To understand how clusters were formed over time (e.g., merged together, etc.)

Comment (MaxHalford) [reply river/cluster/textclust.py:163]: I understand that we need to identify samples with identifiers. But I'm still a bit doubtful as to why this would have to be handled by the clustering model. Can't we just store the cluster assignments outside the clustering model? Indeed, it sounds like a feature you would want to do for any clustering task.

> So why do we need IDs for clusters? To understand how clusters were formed over time (e.g., merged together, etc.)

That sounds like it should be handled internally.

Comment (Dennis1989) [reply river/cluster/textclust.py:163]: > Can't we just store the cluster assignments outside the clustering model?

Yes, we could do this. It would be up to the user to ensure storage. If we decide on this, we should provide some copy&paste examples as I think this is something that people want to do frequently.

Comment (MaxHalford) [reply river/cluster/textclust.py:163]: Yes it's a good excuse to document it properly. Something in the user guide would be nice. You can kill two birds with one stone when you work on the notebook in a future PR :).

I'll let you remove this part from the code, then we can merge and move on.

**Pull request merged**
